pkg/cmd/openshift-operators/apis/operators/v1alpha1/types.go

```diff
@@ -0,0 +1,86 @@
+	Unmanaged ManagementState = "Unmanaged"
+	// Disabled means that the operator is actively managing its resources and trying to remove all traces of the component
+	Disabled ManagementState = "Disabled"
+)
+
+type OperatorSpec struct {
+	// managementState indicates whether and how the operator should manage the component
+	ManagementState ManagementState `json:"managementState"`
+
+	// imagePullSpec is the image to use for the component.
+	ImagePullSpec string `json:"imagePullSpec"`
+
+	// version is the desired state in major.minor.micro-patch.  Usually patch is ignored.
+	Version string `json:"version"`
+
+	// logging contains glog parameters for the component pods.  It's always a command line arg for the moment
+	Logging LoggingConfig `json:"logging,omitempty"`
+}
+
+type LoggingConfig struct {
+	// level is passed to glog.
+	Level int64 `json:"level"`
+
+	// vmodule is passed to glog.
+	Vmodule string `json:"vmodule"`
+}
+
+type ConditionStatus string
+
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+
+	OperatorStatusTypeAvailable = "Available"
+	OperatorStatusTypeMigrating = "Migrating"
+	OperatorStatusTypeFailing   = "Failing"
+)
+
+type OperatorCondition struct {
+	Type               string          `json:"type"`
+	Status             ConditionStatus `json:"status"`
+	LastTransitionTime metav1.Time     `json:"lastTransitionTime,omitempty"`
+	Reason             string          `json:"reason,omitempty"`
+	Message            string          `json:"message,omitempty"`
+}
+
+// VersionAvailablity gives information about the synchronization and operational status of a particular version of the component
+type VersionAvailablity struct {
+	// version is the level this availbility applies to
+	Version string `json:"version"`
+	// updatedReplicas indicates how many replicas are at the desired state
+	UpdatedReplicas int32 `json:"updatedReplicas"`
+	// readyReplicas indicates how many replicas are ready and at the desired state
+	ReadyReplicas int32 `json:"readyReplicas"`
+	// errors indicates what failures are associated with the operator trying to manage this version
+	Errors []string `json:"errors"`
+}
+
+type OperatorStatus struct {
+	// conditions is a list of conditions and their status
+	Conditions []OperatorCondition `json:"conditions,omitempty"`
+
+	// state indicates what the operator has observed to be its current operational status.
+	State ManagementState `json:"state,omitempty"`
+	// task is a high level summary of what the controller is currently attempting to do.  It is high-level, human-readable
+	// and not guaranteed in any way. (I needed this for debugging and realized it made a great summary).
+	Task string `json:"task,omitempty"`
+
+	// currentVersionAvailability is availibility information for the current version
+	CurrentAvailability VersionAvailablity `json:"currentVersionAvailability,omitempty"`
```

Comment:
this should not be omitempty

Reply:
(or made it *VersionAvailability)

Reply:
> this should not be omitempty

When state is disabled, why would a serialization have this?

Reply:
in that case I think you're missing the pointer? (otherwise this will always be serialized as empty struct?)